### PR TITLE
Support usernames in remote url and implement auth for git push

### PR DESCRIPTION
**Pull request by faza**

### CODE REVIEW

File: cmd/git-remote-gitopia/main.go [hunk 1]
```diff
@@ -31,6 +33,25 @@ func Main(args []string, reader io.Reader, writer io.Writer, logger *log.Logger)
...
+			}
+
+			valid, err := regexp.MatchString("^[a-zA-Z0-9]+(?:[-]?[a-zA-Z0-9])*$", remoteUserId)
+			if err != nil {
+				return fmt.Errorf("Invalid remote url")
+			}
+			if !valid {
+				return fmt.Errorf("Invalid remote url")
+			}
+		}
```
Comment (harry-hov) [inline cmd/git-remote-gitopia/main.go:55]: Maybe we can do something like that?
```go
_, err = sdk.AccAddressFromBech32(remoteUserId)
if err != nil {
	if len(remoteUserId) < 3 || len(remoteUserId) > 39 {
		return fmt.Errorf("Invalid remote url")
	}
	valid, err := regexp.MatchString("^[a-zA-Z0-9]+(?:[-]?[a-zA-Z0-9])*$", remoteUserId)
	if err != nil {
		return fmt.Errorf("Invalid remote url")
	}
	if !valid {
		return fmt.Errorf("Invalid remote url")
	}
}
```

File: cmd/git-remote-gitopia/gitopia.go [hunk 2]
```diff
@@ -403,7 +400,17 @@ func (h *GitopiaHandler) Push(remote *core.Remote, refsToPush []core.RefToPush)
 		offchaintypes.RegisterInterfaces(encConf.InterfaceRegistry)
 		offchaintypes.RegisterLegacyAminoCodec(encConf.Amino)
 
-		privKey, err := h.gWallet.privKey()
+		var privKey offchaintypes.SignatureProvider
+
+		if h.secType == KEYRING_BACKEND {
+			// TODO
+			// k, err := sdkkeyring.New(AppName, h.kb.backend, "", os.Stdin)
```
Comment (faza) [inline cmd/git-remote-gitopia/gitopia.go]: @ProgramCpp I want keyring_backend to return some object which supports the following interface

```go
// SignatureProvider offers an abstraction over private keys
// which can be constructed from raw private keys or keyrings
// it is a subset of the methods exposed by cryptotypes.PrivKey
type SignatureProvider interface {
	Sign(msg []byte) (signedBytes []byte, err error)
	PubKey() cryptotypes.PubKey
}
```

I need this to sign request

File: cmd/git-remote-gitopia/gitopia.go [hunk 3]
```diff
@@ -403,7 +400,17 @@ func (h *GitopiaHandler) Push(remote *core.Remote, refsToPush []core.RefToPush)
...
+		} else if h.secType == LEDGER {
+			privKey = h.ledgerPrivateKey
+		} else {
+			privKey, err = h.gWallet.privKey()
+		}
+
 		signer := offchaintypes.NewSigner(encConf.TxConfig, privKey)
```
Comment (faza) [inline cmd/git-remote-gitopia/gitopia.go:457]: @ProgramCpp This is where i am initializing the singer

**Comment by faza** [comment]
@harry-hov This should be reviewed along with the complementary PR in the git-server: https://github.com/gitopia/git-server/pull/30

- git-server expects an auth token in the header when git push request is made (git-receive-pack service in git-server)
- git-server first verifies if the address in the signed data has push permissions (https://github.com/gitopia/git-server/blob/git-server-auth/main.go#L85)
- if yes, then it verifies if the signature is valid

**Review by harry-hov — APPROVED** [review]

**Review by harry-hov — APPROVED** [review]
LGTM

**Pull request merged**